modules/localstack/localstack.go

```diff
@@ -0,0 +1,141 @@
+package localstack
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/internal/testcontainersdocker"
+	"github.com/testcontainers/testcontainers-go/wait"
+	"golang.org/x/mod/semver"
+)
+
+const defaultPort = 4566
+const defaultRegion = "us-east-1"
+const defaultVersion = "0.11.2"
+const hostnameExternalEnvVar = "HOSTNAME_EXTERNAL"
+
+const defaultAccessKeyID = "accesskey"
+const defaultSecretAccessKey = "secretkey"
+const defaultToken = "token"
+
+func isLegacyMode(image string) bool {
+	parts := strings.Split(image, ":")
+	version := parts[len(parts)-1]
+
+	if version == "latest" {
+		return false
+	}
+
+	if !strings.HasPrefix(version, "v") {
+		version = fmt.Sprintf("v%s", version)
+	}
+
+	if semver.IsValid(version) {
+		return semver.Compare(version, "v0.11") < 0 // version < v0.11
+	}
+
+	return true
+}
+
+// StartContainer creates an instance of the LocalStack container type, being possible to pass a custom request and options:
+// - overrideReq: a function that can be used to override the default container request, usually used to set the image version, environment variables for localstack, etc.
+// - opts: a list of functions that can be used to configure the container, more specifically enabling AWS services, or setting the legacy mode.
+func StartContainer(ctx context.Context, overrideReq OverrideContainerRequestOption, opts ...LocalStackContainerOption) (*LocalStackContainer, error) {
+	// defaultContainerRequest {
+	req := testcontainers.ContainerRequest{
+		Image:      fmt.Sprintf("localstack/localstack:%s", defaultVersion),
+		Binds:      []string{fmt.Sprintf("%s:/var/run/docker.sock", testcontainersdocker.ExtractDockerHost(ctx))},
+		WaitingFor: wait.ForLog("Ready.\n").WithOccurrence(1).WithStartupTimeout(2 * time.Minute),
+		Env: map[string]string{
+			"AWS_ACCESS_KEY_ID":     defaultAccessKeyID,
+			"AWS_SECRET_ACCESS_KEY": defaultSecretAccessKey,
+			"AWS_SESSION_TOKEN":     defaultToken,
+			"DEFAULT_REGION":        defaultRegion,
+		},
+	}
+	// }
+
+	localStackReq := LocalStackContainerRequest{
+		ContainerRequest: req,
+	}
+
+	// first, when needed, we merge the user request with the default one
+	if overrideReq != nil {
+		merged := overrideReq(localStackReq.ContainerRequest)
+		localStackReq.ContainerRequest = merged
+	}
+
+	// finally we apply the options
+	for _, opt := range opts {
+		opt(&localStackReq)
+	}
+
+	if isLegacyMode(localStackReq.Image) {
+		return nil, fmt.Errorf("version=%s. Testcontainers for Go does not support running LocalStack in legacy mode. Please use a version >= 0.11.0", localStackReq.Image)
+	}
```

Comment:
If for some reason want to use an older version it's only about to override the ContainerRequest with the right ports which in this case should be manually instead.

Reply:
Indeed, if we want to support legacy mode, then it should not raise an error. My question is: do we want to provide backwards compatibility with legacy mode? If so, I'd not remove the ports from the Service struct, in order to make it easier for developers to switch to legacy.

Reply:
A valid question would also be: do we want to support legacy mode in this brand new module?

Reply:
We agreed in not supporting legacy mode